libs/pilight/protocols/433.92/unitec.c

```diff
@@ -0,0 +1,249 @@
+	 * protocol. See here:
+	 * https://sites.google.com/site/arduinohomeautomation/home/unitec_eim826
+	 * (there 0 and 1 is vice versa, frame bit positions are counted reverse).
+	 */
+
+	for(x=0; x < countof(frame); x++) {
+		// High "1" if first > second, Low "0" else
+		frame[x] = unitec_switch->raw[x*2] > unitec_switch->raw[x*2+1];
+	}
+
+	int idn = binToDec(frame, 0, 19);
+	int unit = binToDec(frame, 20 + 0, 20 + 2) ^ 7;
+	int state = frame[20 + 3];
+
+	unit = unit == 4 ? 0 : (unit + 1);
+
+	char id[16];
+	sprintf(id, "X%X", idn);
+
+	createMessage(id, unit, state);
+}
+
+static void createLow(int s, int c) {
+
+	while(--c >= 0) {
+		int a = 2*s;
+		int b = 2*s++ + 1;
+		unitec_switch->raw[a]=MIN_PULSE_LENGTH;
+		unitec_switch->raw[b]=MAX_PULSE_LENGTH;
+	}
+}
+
+static void createHigh(int s, int c) {
+
+	while(--c >= 0) {
+		int a = 2*s;
+		int b = 2*s++ + 1;
+		unitec_switch->raw[a]=MAX_PULSE_LENGTH;
+		unitec_switch->raw[b]=MIN_PULSE_LENGTH;
+	}
+}
+
+static void clearCode(void) {
+	createLow(0, 20 + 3 + 1);
+}
+
+static void createId(int idn) {
+
+	int pos = 0, len = 20;
+	for(;--len >= 0; idn >>= 1) {
+	    ((idn & 1) ? createHigh : createLow)(pos++, 1);
+	}
+}
+
+static void createUnit(int unit) {
+
+	unit = unit == 0 ? 4 : (unit - 1);
+
+	int pos = 20, len = 3;
+	for(; --len >= 0; unit >>= 1) {
+	    // inverted:
+	    ((unit & 1) ? createLow : createHigh)(pos++, 1);
+	}
+}
+
+static void createState(int state) {
+	// pos = 23, len = 1
+	(state == 0 ? createLow : createHigh)(23, 1);
+}
+
+static void createFooter(void) {
+	unitec_switch->raw[RAW_LENGTH - 2] = MIN_PULSE_LENGTH;
+	unitec_switch->raw[RAW_LENGTH - 1] = OUR_PULSE_DIV;
+	unitec_switch->rawlen = RAW_LENGTH;
+}
+
+static int createCode(const struct JsonNode *code) {
+	const char *id = NULL;
+	int unit = -1;
+	int state = -1;
+	double itmp;
+
+	if(json_find_string(code, "id", &id) != 0) {
+		logprintf(LOG_ERR, "unitec_switch: missing --id argument");
+		return EXIT_FAILURE;
+	}
+
+	if(json_find_number(code, "off", &itmp) == 0)
+		state=0;
+	else if(json_find_number(code, "on", &itmp) == 0)
+		state=1;
+	else {
+		logprintf(LOG_ERR, "unitec_switch: missing state argument (--on or --off)");
+		return EXIT_FAILURE;
+	}
+
+	if(json_find_number(code, "unit", &itmp) == 0) {
+		unit = (int)itmp;
+	} else {
+		logprintf(LOG_ERR, "unitec_switch: missing --unit argument (1..4 or 0 for all)");
+		return EXIT_FAILURE;
+	}
+
+	if(id[0] != 'X' && id[0] != 'x') {
+		logprintf(LOG_ERR, "unitec_switch: id must start with 'X' followed by hex-digits (0..9 and A-F).");
+		return EXIT_FAILURE;
+	}
+	char *end = NULL;
+	int idn = strtol(id+1, &end, 16);
+	if(!end || end==id+1 || *end) {
+		logprintf(LOG_ERR, "unitec_switch: invalid char in id: %s ", end ? end : "?");
+		return EXIT_FAILURE;
+	}
+	if(end -(id+1) > 5) {
+		logprintf(LOG_ERR, "unitec_switch: too long id. Max 5 digits allowed.");
+		return EXIT_FAILURE;
+	}
+
+	if(unit < 0 || unit > 4) {
```

Comment:
The range of unit is 0 .. 4.

Text in line 191 says 1..4 **or 0 for all** which is 0..4 as checked in line 190.

The regex in line 230 is a typo. Will fix that, thank you.